homeassistant/components/ld2410_ble/__init__.py

```diff
@@ -0,0 +1,119 @@
+        )
+    ld2410_ble = LD2410BLE(ble_device)
+
+    coordinator = LD2410BLECoordinator(hass, ld2410_ble)
+
+    try:
+        await ld2410_ble.initialise()
+    except BleakError as exc:
+        raise ConfigEntryNotReady(
+            f"Could not initialise LD2410B device with address {address}"
+        ) from exc
+
+    @callback
+    def _async_update_ble(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        ld2410_ble.set_ble_device_and_advertisement_data(
+            service_info.device, service_info.advertisement
+        )
+
+    entry.async_on_unload(
+        bluetooth.async_register_callback(
+            hass,
+            _async_update_ble,
+            BluetoothCallbackMatcher({ADDRESS: address}),
+            bluetooth.BluetoothScanningMode.ACTIVE,
+        )
+    )
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = LD2410BLEData(
+        entry.title, ld2410_ble, coordinator
+    )
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
+
+    async def _async_stop(event: Event) -> None:
+        """Close the connection."""
+        await ld2410_ble.stop()
+
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_stop)
+    )
+    return True
+
+
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    data: LD2410BLEData = hass.data[DOMAIN][entry.entry_id]
+    if entry.title != data.title:
+        await hass.config_entries.async_reload(entry.entry_id)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        data: LD2410BLEData = hass.data[DOMAIN].pop(entry.entry_id)
+        await data.device.stop()
+
+    return unload_ok
+
+
+class LD2410BLECoordinator(DataUpdateCoordinator):
+    """Data coordinator for receiving LD2410B updates."""
+
+    def __init__(self, hass: HomeAssistant, ld2410_ble: LD2410BLE) -> None:
+        """Initialise the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+        )
+        self._ld2410_ble = ld2410_ble
+        ld2410_ble.register_callback(self._async_handle_update)
+        ld2410_ble.register_disconnected_callback(self._async_handle_disconnect)
+
+    def _async_handle_update(self, state: LD2410BLEState) -> None:
+        """Just trigger the callbacks."""
+        self.async_set_updated_data(True)
+
+    def _async_handle_disconnect(self) -> None:
+        """Trigger the callbacks with False for disconnected."""
+        self.async_set_update_error(
+            Exception(f"Disconnected from {self._ld2410_ble.name}")
+        )
```

Comment:
This isn't really a bluetooth or device exception. `async_set_update_error` takes an exception as a parameter, but in the disconnect callback (which is not exceptional) there is no exception, so I constructed this to pacify the call.

Reply:
You could have the lib pass back a `LD2410DisconnectedError` via the `register_disconnected_callback` and pass that back to `async_set_update_error` so if you ever wanted change the message or raise a different exception based on why the disconnect happened it would be a lib update and no changes in core

Reply:
There's a bit of a code smell there with the tail wagging the dog. The library never had an exception, so why should it construct one for this use case? At best, it should perhaps pass the disconnect reason to the callback, but the exception still needs to be constructed from it at the point of the `async_set_update_error` call.

Reply:
You could instead get rid of the call to `async_set_update_error`, call `async_update_listeners` instead, add a new property to the coordinator called `connected` and override `available` in the entity to check `super().available and self.connected`

Reply:
That's what I had originally written, but I couldn't get it working. I just rewrote it back in, and it now works... ¯\\\_(ツ)\_/¯